### PR TITLE
feat: td goal command

### PR DESCRIPTION
**Pull request by scottlovegrove**
## Summary

Brings the `td goal` command family from `next` to `main`, alongside the SDK upgrade that made it production-ready.

- New `td goal` command with subcommands: `view`, `list`, `create`, `update`, `delete`, `complete`, `uncomplete`, `link`, `unlink`
- Bump `@doist/todoist-sdk` to `10.2.0` (prod release containing the goals API)
- Loosen version regex assertions in `usage-tracking.test.ts` to tolerate prerelease suffixes

## Test plan

- [ ] `npm run type-check` passes
- [ ] `npm test` passes (1582 tests)
- [ ] `td goal --help` shows all subcommands
- [ ] `td goal list` returns goals for the authenticated user
- [ ] `td goal create`, `update`, `complete`, `delete` round-trip
- [ ] `td goal link` / `unlink` associate tasks correctly

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/commands/goal.ts [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+    deadline?: string
+    responsible?: string
+    json?: boolean
+    dryRun?: boolean
+}
+
+async function updateGoal(ref: string, options: UpdateOptions): Promise<void> {
+    const { name, description, deadline, responsible, json, dryRun } = options
+
+    if (!name && !description && !deadline && !responsible) {
```
Comment (doistbot) [inline src/commands/goal.ts]: [P1] Checking options with truthiness (`!name`) prevents users from explicitly unsetting fields using empty strings (e.g., `--description ""`). Empty strings will incorrectly trigger the "No update fields specified" error or be silently ignored when building the update payload.

Use explicit `undefined` checks instead:
```typescript
if (name === undefined && description === undefined && deadline === undefined && responsible === undefined) {
```
And similarly for assigning `args` below (lines 251-254):
```typescript
if (name !== undefined) args.name = name
```

**Review by doistbot — COMMENTED** [review]
This PR successfully introduces the comprehensive `td goal` command family to the main branch alongside the necessary SDK upgrades, significantly expanding the CLI's capabilities for goal management. There are a few areas to refine before merging, including fixing an unawaited promise in task rendering, adding the new read endpoints to the safe methods list, and ensuring proper handling of `CliError` and empty-string updates. Additionally, a few minor adjustments—such as supporting the `--quiet` flag, running network calls concurrently, expanding test coverage, avoiding redundant JSON parsing, and updating `CODEBASE.md`—will help ensure the feature works reliably and smoothly across all edge cases.

<sub>[Share Feedback](https://github.com/Doist/doistbot-app/issues/new?template=review-feedback.yml&title=Review+feedback+for+Doist%2Ftodoist-cli&pr_url=https%3A%2F%2Fgithub.com%2FDoist%2Ftodoist-cli%2Fpull%2F348&datadog_url=https%3A%2F%2Fapp.datadoghq.com%2Flogs%3Fquery%3Dservice%253A%2528doistbot-app-sandbox%2520OR%2520doistbot-app%2529%2520%2540meta.deliveryId%253A6b9c44e0-52ce-11f1-8249-edee0c173da2%26agg_m%3Dcount%26agg_m_source%3Dbase%26agg_t%3Dcount%26clustering_pattern_field_path%3Dmessage%26cols%3Dhost%252Cservice%26fromUser%3Dtrue%26messageDisplay%3Dinline%26refresh_mode%3Dsliding%26storage%3Dhot%26stream_sort%3Ddesc%26viz%3Dstream%26live%3Dtrue) • [Review Logs](https://app.datadoghq.com/logs?query=service%3A%28doistbot-app-sandbox%20OR%20doistbot-app%29%20%40meta.deliveryId%3A6b9c44e0-52ce-11f1-8249-edee0c173da2&agg_m=count&agg_m_source=base&agg_t=count&clustering_pattern_field_path=message&cols=host%2Cservice&fromUser=true&messageDisplay=inline&refresh_mode=sliding&storage=hot&stream_sort=desc&viz=stream&live=true)</sub>

File: src/commands/goal.ts [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+        throw new Error(
```
Comment (doistbot) [inline src/commands/goal.ts]: [P1] `AGENTS.md` requires throwing `CliError` from `src/lib/errors.ts` for all user-facing errors to ensure proper formatting (especially in JSON mode). Using a standard `Error` results in an unhandled exception stack trace.

```typescript
import { CliError } from '../lib/errors.js'
// ...
throw new CliError('invalid_argument', 'No update fields specified. Use --name, --description, --deadline, or --responsible.')
```

File: src/commands/goal.test.ts [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        const consoleSpy = vi.spyOn(console, 'log').mockImplementation(() => {})
+
+        await program.parseAsync(['node', 'td', 'goal', 'create', '--name', 'Ship v2', '--dry-run'])
+
+        expect(mockApi.addGoal).not.toHaveBeenCalled()
+        expect(consoleSpy).toHaveBeenCalledWith(expect.stringContaining('Would create goal'))
+        consoleSpy.mockRestore()
+    })
+})
+
```
Comment (doistbot) [inline src/commands/goal.test.ts:192]: [P2] `AGENTS.md` requires adding corresponding tests for new features. The PR description lists `goal update` round-tripping in the test plan, but there is no `describe('goal update', ...)` block to cover the update command's logic.

File: src/commands/goal.ts [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+
+    let workspaceId: string | undefined
+    if (options.workspace) {
+        const ws = await resolveWorkspaceRef(options.workspace)
+        workspaceId = String(ws.id)
+    }
+
+    const { results: goals, nextCursor } = await paginate(
+        (cursor, limit) =>
+            api.getGoals({
```
Comment (doistbot) [inline src/commands/goal.ts:45]: [P1] These new goal read endpoints need to be added to the read-only allowlist (`KNOWN_SAFE_API_METHODS`, including `getGoal`/`searchGoals` too). Right now the API proxy treats `getGoals` as mutating, so `td goal list`/`td goal view` will fail for users authenticated with `td auth login --read-only`.

File: src/commands/goal.ts [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+
+    for (const task of tasks) {
+        const assignee = formatAssignee({
+            userId: task.responsibleUid,
+            projectId: task.projectId,
+            projects: projectMap,
+            cache: collaboratorCache,
+        })
+        console.log(
+            formatTaskRow({
```
Comment (doistbot) [inline src/commands/goal.ts]: [P1] `formatTaskRow` is async, so logging it without `await` prints a `Promise` instead of the rendered task row. `td goal view`'s pretty output is broken for any goal that has linked tasks.

File: src/commands/goal.ts [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+        deadline: options.deadline ?? null,
+        responsibleUid: options.responsible ?? null,
+    })
+
+    if (options.json) {
+        console.log(formatJson(goal, 'goal'))
+        return
+    }
+
+    console.log(`Created: ${goal.name}`)
```
Comment (doistbot) [inline src/commands/goal.ts:215]: [P2] The success paths in this file never check `isQuiet()`. That means `td goal create --quiet` prints human-readable text instead of the bare ID, and the other goal mutators still emit success messages, which breaks the documented global `--quiet` behavior.

File: src/commands/goal.ts [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+    }
+    console.log('')
+
+    if (tasks.length === 0) {
+        console.log('No linked tasks.')
+        console.log(formatNextCursorFooter(nextCursor))
+        return
+    }
+
+    const { results: projects } = await api.getProjects()
```
Comment (doistbot) [inline src/commands/goal.ts]: [P2] Missed reuse: You can use the exported `fetchProjects(api)` helper from `src/lib/task-list.ts` instead of manually fetching and mapping projects here.

File: src/lib/refs.ts [hunk 8]
```diff
@@ -339,6 +344,19 @@ export async function resolveProjectId(api: TodoistApi, ref: string): Promise<st
     return project.id
 }
 
+export async function resolveGoalRef(api: TodoistApi, ref: string): Promise<Goal> {
+    return resolveRef(
+        ref,
+        (id) => api.getGoal(id),
+        () =>
+            paginate((cursor, limit) => api.getGoals({ cursor: cursor ?? undefined, limit }), {
+                limit: 500,
```
Comment (doistbot) [inline src/lib/refs.ts]: [P2] Hard-capping goal lookup to 500 results makes `resolveGoalRef()` incomplete once an account has more than 500 accessible goals. Other resolvers either search server-side or fetch the full candidate set; since the SDK already exposes `searchGoals`, using that here (or removing the arbitrary ceiling) would keep name-based goal refs correct at scale.

File: src/commands/goal.ts [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+    const goal = await resolveGoalRef(api, ref)
+    await api.uncompleteGoal(goal.id)
+    console.log(`Reopened: ${goal.name}`)
+}
+
+// ── Link / Unlink ──
+
+async function linkGoal(ref: string, options: { task: string }): Promise<void> {
+    const api = await getApi()
+    const goal = await resolveGoalRef(api, ref)
```
Comment (doistbot) [inline src/commands/goal.ts]: [P2] `resolveGoalRef` and `resolveTaskRef` are independent network operations. Consider resolving them concurrently with `Promise.all` to reduce latency.

```typescript
const [goal, task] = await Promise.all([
    resolveGoalRef(api, ref),
    resolveTaskRef(api, options.task),
])
```

File: src/commands/goal.ts [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+          ? parseInt(options.limit, 10)
+          : LIMITS.tasks
+
+    const { results: tasks, nextCursor } = await paginate(
+        (cursor, limit) => api.getTasks({ goalId: goal.id, cursor: cursor ?? undefined, limit }),
+        { limit: targetLimit },
+    )
+
+    if (options.json) {
+        const goalJson = JSON.parse(formatJson(goal, 'goal', options.full))
```
Comment (doistbot) [inline src/commands/goal.ts]: [P2] `goal view --json` stringifies `goal` and `tasks`, parses both back into objects, then stringifies the combined payload again. With the default 300-task page size this adds two extra full traversals of the response body on every JSON view. Please build the combined object from an object-level formatter instead of round-tripping through JSON strings.

File: src/commands/goal.test.ts [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+        mockApi = createMockApi()
+        mockGetApi.mockResolvedValue(mockApi)
+    })
+
+    it('shows goal details and linked tasks', async () => {
+        const program = createProgram()
+        const consoleSpy = vi.spyOn(console, 'log').mockImplementation(() => {})
+
+        mockApi.getGoal.mockResolvedValue(fixtures.goals.shipV2)
+        mockApi.getTasks.mockResolvedValue({ results: [], nextCursor: null })
```
Comment (doistbot) [inline src/commands/goal.test.ts]: [P2] This test claims to cover linked tasks, but `getTasks` is mocked to return an empty list, so the branch that loads projects/collaborators and renders task rows never runs. Add at least one linked task fixture here and assert its rendered output so the main `goal view` path is actually covered.

File: src/index.ts [hunk 12]
```diff
@@ -145,6 +145,7 @@ const commands: Record<string, [string, () => Promise<(p: Command) => void>]> =
         'Manage filters',
         async () => (await import('./commands/filter/index.js')).registerFilterCommand,
     ],
+    goal: ['Manage goals', async () => (await import('./commands/goal.js')).registerGoalCommand],
```
Comment (doistbot) [inline src/index.ts:148]: [P2] This adds a new top-level command, but `CODEBASE.md` was left unchanged. The repo guidance calls out command-registration changes as a trigger to keep that map in sync, so please update `CODEBASE.md` to include the new `goal` command (and any newly exposed goal helpers).

File: src/commands/goal.ts (hunk 1 above)
Comment (scottlovegrove) [reply src/commands/goal.ts]: Fixed in 2484b59 — switched to `=== undefined` checks for both the gate and the args object.

File: src/commands/goal.ts (hunk 2 above)
Comment (scottlovegrove) [reply src/commands/goal.ts]: Fixed in 2484b59 — now throws `CliError('INVALID_ARGUMENT', ...)`.

File: src/commands/goal.test.ts (hunk 3 above)
Comment (scottlovegrove) [reply src/commands/goal.test.ts:192]: Added a `goal update` describe block in 2484b59 covering name update, empty-string clear, missing-fields CliError, dry-run, and `--json`.

File: src/commands/goal.ts (hunk 4 above)
Comment (scottlovegrove) [reply src/commands/goal.ts:45]: Added `getGoals`, `getGoal`, `searchGoals` to `KNOWN_SAFE_API_METHODS` in 2484b59.

File: src/commands/goal.ts (hunk 5 above)
Comment (scottlovegrove) [reply src/commands/goal.ts]: Fixed in 2484b59 — `formatTaskRow` is now awaited, and the linked-task render path is covered by the updated `goal view` test.

File: src/commands/goal.ts (hunk 6 above)
Comment (scottlovegrove) [reply src/commands/goal.ts:215]: Added `isQuiet()` gates to create / update / delete / complete / uncomplete / link / unlink in 2484b59. `create --quiet` now emits just the bare ID.

File: src/commands/goal.ts (hunk 7 above)
Comment (scottlovegrove) [reply src/commands/goal.ts]: Switched to `fetchProjects(api)` from `task-list.ts` in 2484b59.

File: src/lib/refs.ts (hunk 8 above)
Comment (scottlovegrove) [reply src/lib/refs.ts]: Switched `resolveGoalRef` to `api.searchGoals` in 2484b59 — same shape as `resolveTaskRef`, no client-side cap.

File: src/commands/goal.ts (hunk 9 above)
Comment (scottlovegrove) [reply src/commands/goal.ts]: Wrapped both ref resolutions in `Promise.all` for `link` and `unlink` in 2484b59.

File: src/commands/goal.ts (hunk 10 above)
Comment (scottlovegrove) [reply src/commands/goal.ts]: Refactored in 2484b59 — exported `processJsonItem` from `output.ts` and use it to build the composite payload directly, no stringify+parse round-trip.

File: src/commands/goal.test.ts (hunk 11 above)
Comment (scottlovegrove) [reply src/commands/goal.test.ts]: The `goal view` test now mocks `getTasks` with `fixtures.tasks.basic` and `getProjects` with `fixtures.projects.work`, asserts the task content renders, and explicitly checks the output doesn't contain `[object Promise]` (regression guard for the missing-await bug).

File: src/index.ts (hunk 12 above)
Comment (scottlovegrove) [reply src/index.ts:148]: Added the `goal` command to the catalog + tree, listed `resolveGoalRef` under `refs.ts`, and documented the new `processJsonItem` helper under `output.ts` in 2484b59.

**Comment by craigcarlyle** [comment]
✅
